src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -2030,26 +2055,39 @@ bool StatefulWriter::ack_timer_expired()
 
     while (interval.count() < 0)
     {
+        bool acks_flag = false;
         for_matched_readers(matched_local_readers_, matched_datasharing_readers_, matched_remote_readers_,
-                [this](ReaderProxy* reader)
+                [this, &acks_flag](ReaderProxy* reader)
                 {
                     if (reader->disable_positive_acks())
                     {
                         reader->acked_changes_set(last_sequence_number_ + 1);
+                        acks_flag = true;
                     }
                     return false;
                 }
                 );
-        last_sequence_number_++;
+        if (acks_flag)
+        {
+            check_acked_status();
+        }
 
-        // Get the next cache change from the history
         CacheChange_t* change;
 
+        do
+        {
+            last_sequence_number_++;
+        } while (!mp_history->get_change(
+            last_sequence_number_,
+            getGuid(),
+            &change) && last_sequence_number_ < next_sequence_number());
```

Comment:
It would be a good idea to leave a comment here explaining a little bit the intention here. Something like "Skip removed changes until reaching the next change below the next_sequence_number, if possible" What do you think ?